lib/operations/create_index.js

```diff
@@ -1,166 +1,130 @@
-  });
-}
-
-/**
- * Create index using the createIndexes command.
- *
- * @param {Db} db The Db instance on which to execute the command.
- * @param {string} name Name of the collection to create the index on.
- * @param {(string|object)} fieldOrSpec Defines the index.
- * @param {Object} [options] Optional settings. See Db.prototype.createIndex for a list of options.
- * @param {Db~resultCallback} [callback] The command result callback.
- */
-function createIndexUsingCreateIndexes(db, name, fieldOrSpec, options, callback) {
-  // Build the index
-  const indexParameters = parseIndexOptions(fieldOrSpec);
-  // Generate the index name
-  const indexName = typeof options.name === 'string' ? options.name : indexParameters.name;
-  // Set up the index
-  const indexes = [{ name: indexName, key: indexParameters.fieldHash }];
-  // merge all the options
-  const keysToOmit = Object.keys(indexes[0]).concat([
-    'writeConcern',
-    'w',
-    'wtimeout',
-    'j',
-    'fsync',
-    'readPreference',
-    'session'
-  ]);
-
-  for (let optionName in options) {
-    if (keysToOmit.indexOf(optionName) === -1) {
-      indexes[0][optionName] = options[optionName];
+    // Ensure we have a callback
+    if (options.writeConcern && typeof callback !== 'function') {
+      throw MongoError.create({
+        message: 'Cannot use a writeConcern without a provided callback',
+        driver: true
+      });
     }
-  }
 
-  // Get capabilities
-  const capabilities = db.s.topology.capabilities();
+    // Attempt to run using createIndexes command
+    super.execute(cmd, (err, result) => {
+      if (err == null) return handleCallback(callback, err, indexes.name);
+
+      /**
+       * The following errors mean that the server recognized `createIndex` as a command so we don't need to fallback to an insert:
+       * 67 = 'CannotCreateIndex' (malformed index options)
+       * 85 = 'IndexOptionsConflict' (index already exists with different options)
+       * 86 = 'IndexKeySpecsConflict' (index already exists with the same name)
+       * 11000 = 'DuplicateKey' (couldn't build unique index because of dupes)
+       * 11600 = 'InterruptedAtShutdown' (interrupted at shutdown)
+       * 197 = 'InvalidIndexSpecificationOption' (`_id` with `background: true`)
+       */
+      if (
+        err.code === 67 ||
+        err.code === 11000 ||
+        err.code === 85 ||
+        err.code === 86 ||
+        err.code === 11600 ||
+        err.code === 197
+      ) {
+        return handleCallback(callback, err, result);
+      }
 
-  // Did the user pass in a collation, check if our write server supports it
-  if (indexes[0].collation && capabilities && !capabilities.commandsTakeCollation) {
-    // Create a new error
-    const error = new MongoError('server/primary/mongos does not support collation');
-    error.code = 67;
-    // Return the error
-    return callback(error);
+      if (result && result.ok === 0) return handleCallback(callback, toError(result), null);
```

Comment:
`createIndexes` is supported in 2.6 and greater, can we get rid of the fallback behavior?

Reply:
sgtm! is that okay with you @mbroadst?

Reply:
yep!